prompt2model/model_trainer/generate.py

```diff
@@ -236,13 +247,164 @@ def tokenize_dataset(
         }
         return datasets.Dataset.from_dict(preprocessed_dict)
 
+    def search_best_hyperparameters(
+        self,
+        hyperparameters: dict[str, Any],
+        training_datasets: Dataset,
+        validation_datasets: Dataset,
+    ) -> transformers.Trainer:
+        """Method to do automatic hyperparameter search.
+
+        The hyperparameters in this argument will be the same as the hyperparameter
+        choices in the `train_model()` method. However the hyperparameter_choices
+        parameter should follow this schema as shown in this below example.
+
+        ```python
+        hyperparameter_choice = {
+            "static_hyperparameters": {
+                "output_dir": "./result",
+                "logging_steps": 1,
+                "save_strategy": "no",
+                "num_train_epochs": 10,
+                "per_device_train_batch_size": 100,
+                "warmup_steps": 0,
+                "weight_decay": 0.01,
+                "logging_dir": "./result",
+                "learning_rate": 1e-4,
+                "evaluation_strategy": "epoch",
+                "test_size": 0.15,
+            },
+
+            "optuna": {
+                "min_num_train_epochs": 5,
+                "max_num_train_epochs": 10,
+                "save_strategy": ["epoch", "steps", "no"],
+                "evaluation_strategy": ["epoch", "no"],
+                "per_device_train_batch_size": [4, 8, 16, 32],
+                "min_weight_decay": 1e-5,
+                "max_weight_decay": 1e-1,
+                "min_learning_rate": 1e-5,
+                "max_learning_rate": 1e-1,
+            },
+        }
+        ```
+        Here all of the keys are optional. Here are the criterions laid out:
```

Comment:
> I don't think this is entirely accurate, since we're also adding the `optuna` parameters, which are otherwise not given to the `train_model` method. Can you clarify this in the comment?

I have mentioned the schema here, User have two choices, they can either go for hyperpartam optimization or they might not. Now here are some of my questions

`static_hyperparameters` are those which are only meant for model training `trainer.train_model()`.

`optuna` when this key is given, the user can have two choices to put values `default` (this means the hp space will be chosen from the default set) or the user can provide the dict where they can provide all the parameters or some selected ones. Now these will be used during the time of optimizing the hyperparams. And the best hyperparameters values will override the existing `static_hyperparameters`